src/net_processing.cpp

```diff
@@ -5781,30 +5739,35 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
                         if (!txinfo.tx) {
                             continue;
                         }
-                        auto txid = txinfo.tx->GetHash();
                         // Peer told you to not send transactions at that feerate? Don't bother sending it.
                         if (txinfo.fee < filterrate.GetFee(txinfo.vsize)) {
                             continue;
                         }
                         if (tx_relay->m_bloom_filter && !tx_relay->m_bloom_filter->IsRelevantAndUpdate(*txinfo.tx)) continue;
                         // Send
-                        tx_relay->m_recently_announced_invs.insert(hash);
                         vInv.push_back(inv);
                         nRelayedTransactions++;
                         if (vInv.size() == MAX_INV_SZ) {
                             m_connman.PushMessage(pto, msgMaker.Make(NetMsgType::INV, vInv));
                             vInv.clear();
                         }
                         tx_relay->m_tx_inventory_known_filter.insert(hash);
-                        if (hash != txid) {
+                        if (peer->m_wtxid_relay && hash != txinfo.tx->GetHash()) {
```

Comment:
For wtxid relay peers the prior `.insert(hash)` line adds the wtxid to the filter; for non-wtxid relay peers, it adds the txid to the filter. We only need to do a second insert to add the txid for wtxid relay peers, and then only when the wtxid is actually different to the txid. The extra check is just because testing a bool is easier than comparing two uint256s.

Reply:
Is adding the txid still necessary though?

IIUC, the purpose of adding this transaction's txid to `inventory_known_filter` is to save a step if this transaction has an unconfirmed child (if it already existed in this filter, then we wouldn't add it to the `recently_announced` filter). The PR is deleting those lines from `ProcessGetData`, and that's fine since an unconfirmed parent's sequence number should before that of its child.
But I'm not able to see another reason why we should add the txid to the `inventory_known_filter`, so why not delete this whole block?

If it's still needed, it probably makes sense to update the 2 comments since they mention recently_announced filter.

Reply:
Maybe we could recover that as an invariant by having txs added in `MaybeUpdateMempoolForReorg` set the sequence number as 0? The idea being that if they've made it into a block then we've probably announced the headers already so it's no secret that we knew about the tx anyway...

Reply:
> But I'm not able to see another reason why we should add the txid to the `inventory_known_filter`, so why not delete this whole block?

That seems correct to me.

Reply:
Updated to not add txids to the known filter for wtxid peers, and to set the entry sequence to 0 when adding txs that were in a block that was reorged out.

Reply:
This is done in the following commit [c88deec](https://github.com/bitcoin/bitcoin/pull/27675/commits/c88deec99d53f55ff76fedaa30c318be6b72d085)